cmd/cronosd/cmd/changeset.go

```diff
@@ -0,0 +1,687 @@
+	"runtime"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/cosmos/iavl"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/linxGnu/grocksdb"
+	"github.com/spf13/cast"
+	"github.com/spf13/cobra"
+	dbm "github.com/tendermint/tm-db"
+	"github.com/tidwall/btree"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	TimestampSize = 8
+
+	flagStartVersion     = "start-version"
+	flagEndVersion       = "end-version"
+	flagOutput           = "output"
+	flagConcurrency      = "concurrency"
+	flagNoParseChangeset = "no-parse-changeset"
+	flagChunkSize        = "chunk-size"
+	flagZlibLevel        = "zlib-level"
+	flagBatchSize        = "batch-size"
+	flagMoveFiles        = "move-files"
+	flagStore            = "store"
+
+	DefaultChunkSize = 1000000
+
+	CompressedFileSuffix = ".zz"
+	SSTFileExtension     = ".sst"
+
+	StorePrefixTmp = "s/k:%s/"
+)
+
+func ChangeSetGroupCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "changeset",
+		Short: "dump changeset from iavl versions or manage changeset files",
+	}
+	cmd.AddCommand(
+		DumpFileChangeSetCmd(),
+		PrintPlainFileCmd(),
+		ConvertPlainToSSTTSCmd(),
+		IngestSSTCmd(),
+	)
+	return cmd
+}
+
+func DumpFileChangeSetCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "dump-file [store]",
+		Short: "Extract changeset from iavl versions, and save to plain file format",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := server.GetServerContextFromCmd(cmd)
+			if err := ctx.Viper.BindPFlags(cmd.Flags()); err != nil {
+				return err
+			}
+
+			db, err := openDBReadOnly(ctx.Viper.GetString(flags.FlagHome), server.GetAppDBBackend(ctx.Viper))
+			if err != nil {
+				return err
+			}
+			store := args[0]
+			prefix := []byte(fmt.Sprintf(StorePrefixTmp, store))
+			db = dbm.NewPrefixDB(db, prefix)
+
+			cacheSize := cast.ToInt(ctx.Viper.Get(server.FlagIAVLCacheSize))
+
+			startVersion, err := cmd.Flags().GetInt64(flagStartVersion)
+			if err != nil {
+				return err
+			}
+			endVersion, err := cmd.Flags().GetInt64(flagEndVersion)
+			if err != nil {
+				return err
+			}
+			concurrency, err := cmd.Flags().GetInt(flagConcurrency)
+			if err != nil {
+				return err
+			}
+			chunkSize, err := cmd.Flags().GetInt(flagChunkSize)
+			if err != nil {
+				return err
+			}
+			zlibLevel, err := cmd.Flags().GetInt(flagZlibLevel)
+			if err != nil {
+				return err
+			}
+
+			outDir, err := cmd.Flags().GetString(flagOutput)
+			if err != nil {
+				return err
+			}
+			outDir = filepath.Join(outDir, store)
+			if err := os.MkdirAll(outDir, os.ModePerm); err != nil {
+				return err
+			}
+
+			if endVersion == 0 {
+				tree, err := iavl.NewMutableTree(db, 0, true)
+				if err != nil {
+					return err
+				}
+				latestVersion, err := tree.LazyLoadVersion(0)
+				if err != nil {
+					return err
+				}
+				endVersion = latestVersion + 1
+			}
+
+			tasksChan := make(chan *dumpTask, 2^32)
```

Comment:
it looks like, this meant to be 2 << 32? 2^32==34

Reply:
ah, yes

Reply:
it's reallocate the buffer, so I change it to 2048, should be big enough.